hyperledger/iroha
# [fix] #4267: Introduce p2p idle timeout

## Description

Previously iroha tried to connect indefinitely to idle peer.

Introduce idle timeout for p2p communication:

Each peer track the last time it received message (or ping) from particular peer, if this time is larger than timeout peer is disconnected.

If no other messages were produced in time equal to half timeout ping is send to prove that peer isn't idle.





### Linked issue



Closes #4267 



### How to test

1. setup iroha network with multiple peers using `LOGLEVEL=TRACE ./scripts/test_env.py setup`
2. kill one of the peers (e.g. iroha1)
3. start listening on p2p port this peer was using `nc -l 1338`
4. check that peers terminate connection with this "peer" after timeout
